### Resolución TP1

#### 4. Implemente algún método de búsqueda. Puedes elegir cualquiera menos búsqueda en anchura primero (el desarrollado en clase). Sos libre de elegir cualquiera de los vistos en clases, o inclusive buscar nuevos.

In [1]:
# Paquetes necesarios
from hanoi_states import StatesHanoi    # Para definir los estados
from hanoi_states import ProblemHanoi   # Para definir el problema: Estado inicial --> Estado final
from tree_hanoi import NodeHanoi        # Para poder expandir los nodos
from collections import deque           # Para utilizar deque y agregar | quitar elementos de forma rápida
import tracemalloc                      # Para ver demanda de memoria

#### BFS - Búsqueda primero en anchura (explorando todo los nodos)

In [2]:
        # ----------------Input Discos  ----------------

        # Hay que hacerlo como mucho para 3 discos, por tiempos de ejecución...
        ndisks = 3

        # ----------------Problema de Hanoi ----------------

        initial_state = StatesHanoi(list(range(ndisks, 0, -1)), [], [], max_disks=ndisks)
        goal_state = StatesHanoi([], [], list(range(ndisks, 0, -1)), max_disks=ndisks)

        # Se define el problema : ir de estado inicial al final y movimientos entre estados
        problem = ProblemHanoi(initial= initial_state, goal = goal_state)

# Se define la raíz del árbol para probar los algoritmos de búsqueda
root = NodeHanoi(state=initial_state)


#---------------- Algoritmo de búsqueda ----------------

frontier = [NodeHanoi(problem.initial)]  # Creamos una cola FIFO con el nodo inicial

# Mientras que la cola no este vacia
while len(frontier) != 0:
    node = frontier.pop()  # Extraemos el primer nodo de la cola
    if problem.goal_test(node.state):  # Comprobamos si hemos alcanzado el estado objetivo
        last_node = node
        print("Encontramos la solución")
        break
    
    # Agregamos a la cola todos los nodos sucesores del nodo actual
    for next_node in node.expand(problem):
        frontier.insert(0, next_node)

print(f'Cantidad de movimientos: {last_node.state.accumulated_cost}')

Encontramos la solución
Cantidad de movimientos: 7.0


In [49]:
%%timeit -r 10 -n 1

# ---------------- Tiempos de ejecución ----------------

frontier = [NodeHanoi(problem.initial)]  # Creamos una cola FIFO con el nodo inicial

# Mientras que la cola no este vacia
while len(frontier) != 0:
    node = frontier.pop()  # Extraemos el primer nodo de la cola
    if problem.goal_test(node.state):  # Comprobamos si hemos alcanzado el estado objetivo
        last_node = node
        break
    
    # Agregamos a la cola todos los nodos sucesores del nodo actual
    for next_node in node.expand(problem):
        frontier.insert(0, next_node)


55.2 ms ± 4.22 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)


In [5]:
# ---------------- Memoria Consumida ----------------

tracemalloc.start()

frontier = [NodeHanoi(problem.initial)]  # Creamos una cola FIFO con el nodo inicial

# Mientras que la cola no este vacia
while len(frontier) != 0:
    node = frontier.pop()  # Extraemos el primer nodo de la cola
    if problem.goal_test(node.state):  # Comprobamos si hemos alcanzado el estado objetivo
        last_node = node
        break
    
    # Agregamos a la cola todos los nodos sucesores del nodo actual
    for next_node in node.expand(problem):
        frontier.insert(0, next_node)

_, memory_peak = tracemalloc.get_traced_memory()
memory_peak /= 1024*1024
tracemalloc.stop()

print(f"Maxima memoria ocupada: {round(memory_peak, 2)} [MB]", )

Maxima memoria ocupada: 4.58 [MB]


#### BFS - Búsqueda primero en anchura (explorando nodos únicos)

In [6]:
# ----------------Input Discos  ----------------

ndisks = 5

# ----------------Problema de Hanoi ----------------

# Estados inicial | objetivo
initial_state = StatesHanoi(list(range(ndisks, 0, -1)), [], [], max_disks=ndisks)
goal_state = StatesHanoi([], [], list(range(ndisks, 0, -1)), max_disks=ndisks)

# Se define el problema: ir de estado inicial al final y movimientos entre estados
problem = ProblemHanoi(initial= initial_state, goal = goal_state)

# Se define la raíz del árbol para probar los algoritmos de búsqueda
root = NodeHanoi(state=initial_state)


#---------------- Algoritmo de búsqueda ----------------

frontier = [NodeHanoi(problem.initial)]  # Creamos una cola FIFO con el nodo inicial
explored = set()  # Este set nos permite ver si ya exploramos un estado para evitar repetir indefinidamente

# Mientras que la cola no este vacia
while len(frontier) != 0:
    node = frontier.pop()  # Extraemos el primer nodo de la cola
    
    # Agregamos nodo al set. Esto evita guardar duplicados, porque set nunca tiene elementos repetidos
    explored.add(node.state)
    
    if problem.goal_test(node.state):  # Comprobamos si hemos alcanzado el estado objetivo
        last_node = node
        print("Encontramos la solución")
        break
    
    # Agregamos a la cola todos los nodos sucesores del nodo actual
    for next_node in node.expand(problem):
        # Solo si no fue explorado
        if next_node.state not in explored:
            frontier.insert(0, next_node)

print(f'Cantidad de movimientos: {last_node.state.accumulated_cost}')

Encontramos la solución
Cantidad de movimientos: 31.0


In [7]:
%%timeit -r 10 -n 1

# ---------------- Tiempos de ejecución ----------------

frontier = [NodeHanoi(problem.initial)]  # Creamos una cola FIFO con el nodo inicial
explored = set()  # Este set nos permite ver si ya exploramos un estado para evitar repetir indefinidamente

# Mientras que la cola no este vacia
while len(frontier) != 0:
    node = frontier.pop()  # Extraemos el primer nodo de la cola
    
    # Agregamos nodo al set. Esto evita guardar duplicados, porque set nunca tiene elementos repetidos
    explored.add(node.state)
    
    if problem.goal_test(node.state):  # Comprobamos si hemos alcanzado el estado objetivo
        last_node = node
        break
    
    # Agregamos a la cola todos los nodos sucesores del nodo actual
    for next_node in node.expand(problem):
        # Solo si no fue explorado
        if next_node.state not in explored:
            frontier.insert(0, next_node)

84.1 ms ± 5.26 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)


In [8]:
# ---------------- Memoria Consumida ----------------

tracemalloc.start()

frontier = [NodeHanoi(problem.initial)]  # Creamos una cola FIFO con el nodo inicial
explored = set()  # Este set nos permite ver si ya exploramos un estado para evitar repetir indefinidamente

# Mientras que la cola no este vacia
while len(frontier) != 0:
    node = frontier.pop()  # Extraemos el primer nodo de la cola
    
    # Agregamos nodo al set. Esto evita guardar duplicados, porque set nunca tiene elementos repetidos
    explored.add(node.state)
    
    if problem.goal_test(node.state):  # Comprobamos si hemos alcanzado el estado objetivo
        last_node = node
        break
    
    # Agregamos a la cola todos los nodos sucesores del nodo actual
    for next_node in node.expand(problem):
        # Solo si no fue explorado
        if next_node.state not in explored:
            frontier.insert(0, next_node)

_, memory_peak = tracemalloc.get_traced_memory()
memory_peak /= 1024*1024
tracemalloc.stop()

print(f"Maxima memoria ocupada: {round(memory_peak, 2)} [MB]", )

Maxima memoria ocupada: 1.6 [MB]


#### DFS - Búsqueda primero en profundidad (explorando nodos únicos)

In [9]:
# ----------------Input Discos  ----------------

ndisks = 5

# ----------------Problema de Hanoi ----------------

# Estados inicial | objetivo
initial_state = StatesHanoi(list(range(ndisks, 0, -1)), [], [], max_disks=ndisks)
goal_state = StatesHanoi([], [], list(range(ndisks, 0, -1)), max_disks=ndisks)

# Se define el problema: ir de estado inicial al final y movimientos entre estados
problem = ProblemHanoi(initial= initial_state, goal = goal_state)

# Se define la raíz del árbol para probar los algoritmos de búsqueda
root = NodeHanoi(state=initial_state)


#---------------- Algoritmo de búsqueda ----------------

frontier = [NodeHanoi(problem.initial)]  # Creamos una cola FIFO con el nodo inicial
explored = set()  # Este set nos permite ver si ya exploramos un estado para evitar repetir indefinidamente

# Mientras que la cola no este vacia
while len(frontier) != 0:
    node = frontier.pop()  # Extraemos el primer nodo de la cola
    
    # Agregamos nodo al set. Esto evita guardar duplicados, porque set nunca tiene elementos repetidos
    explored.add(node.state)
    
    if problem.goal_test(node.state):  # Comprobamos si hemos alcanzado el estado objetivo
        last_node = node
        print("Encontramos la solución")
        break
    
    # Agregamos a la cola todos los nodos sucesores del nodo actual
    for next_node in node.expand(problem):
        # Solo si no fue explorado
        if next_node.state not in explored:
            frontier.append(next_node)

print(f'Cantidad de movimientos: {last_node.state.accumulated_cost}')

Encontramos la solución
Cantidad de movimientos: 121.0


In [10]:
%%timeit -r 10 -n 1

# ---------------- Tiempos de ejecución ----------------

frontier = [NodeHanoi(problem.initial)]  # Creamos una cola FIFO con el nodo inicial
explored = set()  # Este set nos permite ver si ya exploramos un estado para evitar repetir indefinidamente

# Mientras que la cola no este vacia
while len(frontier) != 0:
    node = frontier.pop()  # Extraemos el primer nodo de la cola
    
    # Agregamos nodo al set. Esto evita guardar duplicados, porque set nunca tiene elementos repetidos
    explored.add(node.state)
    
    if problem.goal_test(node.state):  # Comprobamos si hemos alcanzado el estado objetivo
        last_node = node
        break
    
    # Agregamos a la cola todos los nodos sucesores del nodo actual
    for next_node in node.expand(problem):
        # Solo si no fue explorado
        if next_node.state not in explored:
            frontier.append(next_node)

7.72 ms ± 1.21 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)


In [11]:
# ---------------- Memoria Consumida ----------------

tracemalloc.start()

frontier = [NodeHanoi(problem.initial)]  # Creamos una cola FIFO con el nodo inicial
explored = set()  # Este set nos permite ver si ya exploramos un estado para evitar repetir indefinidamente

# Mientras que la cola no este vacia
while len(frontier) != 0:
    node = frontier.pop()  # Extraemos el primer nodo de la cola
    
    # Agregamos nodo al set. Esto evita guardar duplicados, porque set nunca tiene elementos repetidos
    explored.add(node.state)
    
    if problem.goal_test(node.state):  # Comprobamos si hemos alcanzado el estado objetivo
        last_node = node
        break
    
    # Agregamos a la cola todos los nodos sucesores del nodo actual
    for next_node in node.expand(problem):
        # Solo si no fue explorado
        if next_node.state not in explored:
            frontier.append(next_node)

_, memory_peak = tracemalloc.get_traced_memory()
memory_peak /= 1024*1024
tracemalloc.stop()

print(f"Maxima memoria ocupada: {round(memory_peak, 2)} [MB]", )

Maxima memoria ocupada: 0.25 [MB]


#### DLS - Búsqueda de profundidad limitada (explorando nodos únicos)  

In [21]:
# ----------------Input Discos  ----------------

ndisks = 5
reached = 2 # cantidad de niveles hasta donde explorar

# ----------------Problema de Hanoi ----------------

# Estados inicial | objetivo
initial_state = StatesHanoi(list(range(ndisks, 0, -1)), [], [], max_disks=ndisks)
goal_state = StatesHanoi([], [], list(range(ndisks, 0, -1)), max_disks=ndisks)

# Se define el problema: ir de estado inicial al final y movimientos entre estados
problem = ProblemHanoi(initial= initial_state, goal = goal_state)

# Se define la raíz del árbol para probar los algoritmos de búsqueda
root = NodeHanoi(state=initial_state)


#---------------- Algoritmo de búsqueda ----------------


frontier = [NodeHanoi(problem.initial)]  
explored = set()  # Uso de set para evitar repetir estados

# Mientras que la cola no este vacia
while len(frontier) != 0:
    node = frontier.pop()  # Extraemos el primer nodo de la cola
    
    # Agregamos nodo al set. Esto evita guardar duplicados, porque set nunca tiene elementos repetidos
    explored.add(node.state)
    
    if problem.goal_test(node.state):  # Comprobamos si hemos alcanzado el estado objetivo
        last_node = node
        print("Encontramos la solución")
        break
    
    # Agregamos a la cola todos los nodos sucesores del nodo actual
    if node.depth < reached:
        for next_node in node.expand(problem):
            # Solo si no fue explorado
            if next_node.state not in explored:
                frontier.append(next_node)

if last_node:
    print(f'Cantidad de movimientos: {last_node.state.accumulated_cost}')
else:
    print(f'No se encontró solución hasta este nivel.')

Cantidad de movimientos: 121.0


#### Eficiencia Algoritmos 

| Algoritmo        | Discos | Movimientos | Tiempo prom. ejecución | Memoria consumida |
|------------------|:------:|:-----------:|:----------------------:|:-----------------:|
| BFS (All)        | 3 (*)  | 7           | 55.2 ms ± 4.22 ms      | 4.58MB            |
| BFS (Unique)     | 5      | 31          | 84.1 ms ± 5.26 ms      | 1.6MB             |
| DFS              | 5      | 121         | 7.72 ms ± 1.21 ms      | 0.25MB            |
| Profundidad lim. | 5      | 19          | 3.0s                   | 148MB             |

(*) Se utilizaron únicamente 3 discos por tiempos de ejecución.